### PR TITLE
Make annotation typechecking lazier

### LINKED ISSUE #11870: 2.13.1 scalac StackOverflowError when annotation argument references the annottee
Consider the example:
```scala

class Invariant(condition: Unit => Boolean) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Invariant( _ => x.length == 3 )
  def x: String = "foo"
}
```

This code results in the `java.lang.StackOverflowError` through `scala.reflect.internal.AnnotationInfos$LazyAnnotationInfo.forcedInfo$lzycompute(AnnotationInfos.scala:158)` ([scalac-SOE-stack.txt](https://github.com/scala/bug/files/4129572/scalac-SOE-stack.txt) has the full stack trace if you need it). As far as I can tell, the stack trace is different from those encountered in #7449, #1632, #1667 and #10421. Increasing the stack `-Xss128m` just makes the SOE stack trace longer, so it looks like cyclic init.

Some pointers to help narrow the problem.

1. The annotation argument does not have to be a lambda. The following snippet yields the same SOE: 
```scala
class Annot(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot(x)
  def x: String = "foo"
}
```

2. The following example compiles fine.
```scala
class Annot(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot(y)
  def x: String = "foo"
  def y: String = "bar"
}
```

3. However, if you put the annotation on the second field as well, there is the SOE again
```scala
class Annot(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot(y)
  def x: String = "foo"

 @Annot(x)
  def y: String = "bar"
}
```

4. Finally, the code below causes SOE even when the annotations are different.
```scala
class Annot1(arg: Any) extends scala.annotation.StaticAnnotation
class Annot2(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot1(y)
  def x: String = "foo"

  @Annot2(x)
  def y: String = "bar"
}
```

Thank you!

### PR DESCRIPTION
**Pull request by retronym**
Make annotation typechecking lazier

Now that `Typer.stabilize` ends up checking for `@uncheckedStable` annotations to support defs's as stable paths, new opportunities arise for cyclic errors.

This PR reduces the likelihood of cycles by typechecking the only core of an annotation (_not_ the full application to the annotation arguments) to determine its type symbol.

pos/unchecked-stable-cyclic-error.scala test case started to fail since #8338 with:

```
|-- object Test BYVALmode-EXPRmode (site: package <empty>)
|    |-- super APPSELmode-EXPRmode-POLYmode-QUALmode (silent: <init> in Test)
|    |    |-- this EXPRmode (silent: <init> in Test)
|    |    |    \-> Test.type
|    |    \-> Test.type
|    |-- def foo BYVALmode-EXPRmode (site: object Test)
|    |    |-- attr EXPRmode (site: method foo in Test)
|    |    |    |-- Int TYPEmode (site: method attr in Test)
|    |    |    |    \-> Int
|    |    |    |-- new anno APPSELmode-BYVALmode-EXPRmode-FUNmode-POLYmode (site: object Test)
|    |    |    |    |-- new anno APPSELmode-EXPRmode-POLYmode-QUALmode (site: object Test)
|    |    |    |    |    |-- anno FUNmode-TYPEmode (site: object Test)
|    |    |    |    |    |    \-> anno
|    |    |    |    |    \-> anno
|    |    |    |    \-> (a: Any): anno
|    |    |    |-- (a: Any): anno : pt=anno EXPRmode (site: value <local Test> in Test)
|    |    |    |    |-- foo : pt=Any BYVALmode-EXPRmode (site: value <local Test> in Test)
|    |    |    |    |    caught scala.reflect.internal.Symbols$CyclicReference: illegal cyclic reference involving method foo: while typing foo
/Users/jz/code/scala/sandbox/test.scala:7: error: recursive method foo needs result type
  @anno(foo) def attr: Int = foo
        ^
|    |    |    |    \-> anno
```

Another variant, pos/annotation-cycle.scala, fails only on 2.12.x (after the backport of #8338). Backporting this fix makes that test start passing on 2.12.x.

References https://github.com/scala/bug/issues/12216 (which will be fixed after this is backported to 2.12.x)

Fixes scala/bug#11870

### CODE REVIEW

**Comment by lrytz** [comment]
@retronym see also https://github.com/scala/scala/pull/9144 where I had some reservations about building more stuff on top of `LazyAnnotationInfo` - though your patch here looks small enough. Can you add the test case for https://github.com/scala/bug/issues/11870 from there?

**Comment by retronym** [comment]
@lrytz Test case added.

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 1]
```diff
@@ -1900,11 +1900,27 @@ trait Namers extends MethodSynthesis {
...
+              tpt1.tpe.finalResultType.typeSymbol
+            }
+            AnnotationInfo.lazily(computeSymbol, computeInfo)
+          case _ =>
+            AnnotationInfo lazily {
+              enteringTyper {
+                val annotSig = newTyper(ctx.makeNonSilent(ann)).typedAnnotation(ann, Some(annotee))
+                if (pred(annotSig)) annotSig else UnmappableAnnotation // UnmappableAnnotation will be dropped in typedValDef and typedDefDef
+              }
+            }
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala]:
```suggestion
            AnnotationInfo.lazily(enteringTyper(computeInfo))
```

**Review by lrytz — APPROVED** [review]
LGTM otherwise. The main differences to https://github.com/scala/scala/pull/9144 are:
  - Here we have `ExtraLazyAnnotationInfo extends LazyAnnotationInfo`, where the other PR did `QualifierTypedAnnotationInfo extends AnnotationInfo`. This requires fewer overrides.
  - Here, typing the qualifier / annotation symbol is lazy, in the other PR it's done eagerly during Namer.

**Pull request merged**